channeldb/kvdb/backend.go

```diff
@@ -21,30 +28,97 @@ func fileExists(path string) bool {
 
 // GetBoltBackend opens (or creates if doesn't exits) a bbolt
 // backed database and returns a kvdb.Backend wrapping it.
-func GetBoltBackend(path, name string, noFreeListSync bool) (Backend, error) {
+func GetBoltBackend(path, name string, noFreeListSync,
+	autoCompact bool) (Backend, error) {
+
 	dbFilePath := filepath.Join(path, name)
-	var (
-		db  Backend
-		err error
-	)
 
+	// Is this a new database?
 	if !fileExists(dbFilePath) {
 		if !fileExists(path) {
 			if err := os.MkdirAll(path, 0700); err != nil {
 				return nil, err
 			}
 		}
 
-		db, err = Create(BoltBackendName, dbFilePath, noFreeListSync)
-	} else {
-		db, err = Open(BoltBackendName, dbFilePath, noFreeListSync)
+		return Create(BoltBackendName, dbFilePath, noFreeListSync)
+	}
+
+	// This is an existing database. We might want to compact it on startup
+	// to free up some space.
+	if autoCompact {
+		if err := compactAndSwap(path, name); err != nil {
+			return nil, err
+		}
+	}
+
+	return Open(BoltBackendName, dbFilePath, noFreeListSync)
+}
+
+// compactAndSwap will attempt to write a new temporary DB file to disk with
+// the compacted database content, then atomically swap (via rename) the old
+// file for the new file by updating the name of the new file to the old.
+func compactAndSwap(path, sourceName string) error {
+	// If the main DB file isn't set, then we can't proceed.
+	if sourceName == "" {
+		return fmt.Errorf("cannot compact DB with empty name")
 	}
 
+	sourceFilePath := filepath.Join(path, sourceName)
+	tempDestFilePath := filepath.Join(path, DefaultTempDBFileName)
+	log.Infof("Compacting database file at %v", sourceFilePath)
+
+	// If the old temporary DB file still exists, then we'll delete it
+	// before proceeding.
+	if _, err := os.Stat(tempDestFilePath); err == nil {
```

Comment:
Comment got lost somehow, but: we should check that the user has enough disk space here to accommodate a worst case compaction (basically no gain at all so 2x disk space).